src/cobra/manipulation/modify.py

```diff
@@ -166,9 +170,13 @@ def rename_genes(model: "Model", rename_dict: Dict[str, str]) -> None:
             # rename old gene to new gene
             gene = model.genes[gene_index]
             # trick DictList into updating index
-            model.genes._dict.pop(gene.id)  # ugh
+            model.genes.remove(gene.id)
             gene.id = new_name
             model.genes[gene_index] = gene
+            if context:
+                model.genes.remove(gene.id)
```

Comment:
Shouldn't those be inside a `context(...)` call. Looks odd that having a context should trigger an immediate removal.

Reply:
You are correct. I tried setting up the context to be context(partial()) call, but I'm not having any luck. If you can review this section, and fix it, I'd appreciate it.
Not sure how to do this correctly.

Reply:
Okay, will take a look.

Reply:
We might have to reset the numeric index in some way, because asking 
```
if 'foo' in model.genes
```
relies on the model.genes._dict, so partially on the index, I think.

Reply:
Yeah I meant only the numeric index. So model.genes[id] == gene has to be true but if it was the 5th entry before it can now be the 21st entry (position does not matter).